### ISSUE TITLE
[Resource / Translation] Sorting by translatable field while listing resources

### ISSUE THREAD

**Issue by pamil** [Sylius/Sylius#3836]
### Expected behaviour

``` gherkin
Feature: Sorting products by name

Background:
    Given there is a product with Polish name "Cebula" and English name "Onion"
    And there is another product with Polish name "Marchew" and English name "Carrot"

Scenario: Sorting products by Polish names
    Given I browse the site using Polish locale
    When I list all products sorted by name
    Then "Cebula" should be before "Marchew"

Scenario: Sorting products by English names
    Given I browse the site using English locale
    When I list all products sorted by name
    Then "Carrot" should be before "Onion"
```
### Actual behaviour

When we're listing a list of any translatable resource that is sorted by translatable field, we cannot be sure in which order they show up (probably depends on the database implementation).

For example, the query for a list of products looks like (in pseudo-DQL): `SELECT DISTINCT p FROM products p LEFT JOIN p.translations t ORDER BY t.name`. If product has more translations that one, the sorting is broken as it may use different locales translations to make the sorting.

Doctrine ORM in 2.5 version provides [`LimitSubqueryWalker::validate()`](https://github.com/doctrine/doctrine2/blob/master/lib/Doctrine/ORM/Tools/Pagination/LimitSubqueryWalker.php#L128) that does not allow us to make those kind of Schrodinger-queries and therefore Sylius fails to support it for now.

**Comment by nakashu** [comment]
Is the way how this should be fixed already discussed somewhere ?
If we could have a proposed solution here, someone could make a PR sooner for this.

**Comment by peteward** [comment]
hmmm yeah that's a tricky issue :/

It's Doctrine breaking a perfectly valid query by adding distinct on it for pagination which then makes it technically invalid... 

Is it possible to call one of the other walkers like `WhereInWalker` or `CountWalker`?

**Comment by nakashu** [comment]
Stumbled upon this bug again. Couldn't we just inject or set the `localProvider` to `ORM\translatableResourceRepository`? 

The same way it is done in `Sylius\Bundle\TranslationBundle\Doctrine\ODM\MongoDB\TranslatableResourceRepository`

```php
  /**
     * {@inheritdoc}
     */
    public function setLocaleProvider(LocaleProviderInterface $localeProvider)
    {
        $this->localeProvider = $localeProvider;

        return $this;
    }
    // ...
/**
     * @param string $name
     *
     * @return string
     */
    protected function getPropertyName($name)
    {
        if (in_array($name, $this->translatableFields)) {
            return 'translations.'.$this->localeProvider->getCurrentLocale().'.'.$name;
        }

        return $name;
    }
```

Then we can set it the locale form provider when creating the query builder

```php
protected function getCollectionQueryBuilder()
    {
        $queryBuilder = parent::getCollectionQueryBuilder();

        $queryBuilder
            ->addSelect('translation')
            ->leftJoin($this->getAlias().'.translations', 'translation')
            ->andWhere('translation.locale = :locale')->setParameter(':locale', $this->localeProvider->getCurrentLocale())
        ;

        return $queryBuilder;
    }
```

**Comment by pjedrzejewski** [comment]
Repository cannot depend on this state of the application - locale. We should pass it as argument.

**Comment by nakashu** [comment]
@pjedrzejewski Just to be clear, when you say pass as argument you mean:

1. to the setter method - eg. with kernel listener
2. to find/create methods on repository: eg.: findByStuff($criteria) // and have it in criteria

btw could this be addressed in the merge of `ResoureBundle` and `TranslationBundle` or separatelly after merge

**Comment by michalmarcinkowski** [comment]
The `getCollectionQueryBuilder` and similar methods will be removed [here](https://github.com/Sylius/Sylius/pull/3854). You should pass `locale` explicitly to repository method instead of using provider inside repository.

**Comment by nakashu** [comment]
thanks michal

**Comment by steffenbrem** [comment]
Are there any new updates on this? I really need to use the Doctrine `2.5` since there are some important bug fixes in there for me (such as `postLoad` event etc.), currently I don't see a way how to solve this easily but maybe someone else has some ideas on how to solve this?

@ocramius do you maybe know why it is failing on doctrine 2.5? (Linked issue: https://github.com/Sylius/Sylius/issues/3268)

**Comment by pamil** [comment]
@NeverResponse can you confirm that this issue is solved by your #6180?
